### ISSUE TITLE
Errors after upgrading to 5.2.0 from 5.1.5 on Linux

### ISSUE THREAD

**Issue by alex-jitbit** [dotnet/SqlClient#2378]
After upgrading 5.1.5 to 5.2.0 on Linux (Ubuntu) .NET 8 I'm getting thousands of errors:
```
Exception message:

Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)
 ---> System.TimeoutException: The socket couldn't connect during the expected 14965 remaining time.

Stack trace:

Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)
 ---> System.TimeoutException: The socket couldn't connect during the expected 14965 remaining time.
   at Microsoft.Data.SqlClient.SNI.SNITCPHandle.Connect(String serverName, Int32 port, TimeoutTimer timeout, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo)
   at Microsoft.Data.SqlClient.SNI.SNITCPHandle..ctor(String serverName, Int32 port, TimeoutTimer timeout, Boolean parallel, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo, Boolean tlsFirst, String hostNameInCertificate, String serverCertificateFilename)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry, SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open(SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open()
```

Some requests work fine, but about 50% throw this error. Reverting back to 5.1.5 solves the problem.

### Further technical details
Microsoft.Data.SqlClient version: 5.2
.NET target: NET 8
SQL Server version: SQL 2017 on Linux
Operating system: Ubuntu 22

**Comment by alex-jitbit** [comment]
I suspect this has something to do with your SNI "improvements" in 5.2.0 https://github.com/dotnet/SqlClient/releases/tag/V5.2.0

**Comment by JRahnama** [comment]
@alex-jitbit is this happening on a regular connection? I mean there is not AAD included? can yu provide a sample repro please?

**Comment by JRahnama** [comment]
Linux uses managed SNI and I think the improvements were done mostly on the native side. Which change did you mean?

**Comment by JRahnama** [comment]
@alex-jitbit I will test it today and will update you after.

**Comment by guiestimoneon** [comment]
I had the same problem, I downgrade to 5.1.5 and it worked again

SQL 2019 - Windows Server - .NET 8

**Comment by JRahnama** [comment]
I was not able to repro the issue on Ubuntu 22.04 as a local server, but I will test it with a remote server. If there is any issue it should be related to https://github.com/dotnet/SqlClient/pull/1029

**Comment by JRahnama** [comment]
here is my test setup:

csproj:
```C#
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Data.SqlClient" Version="5.2.0" />
  </ItemGroup>
</Project>
```
Program.cs
```C#
using Microsoft.Data.SqlClient;

SqlConnectionStringBuilder builder = new(){
    DataSource ="*******.database.windows.net",
    UserID = "******",
    Password = "*****",
    InitialCatalog = "Northwind",
    MaxPoolSize = 250
};
using SqlConnection conn = new(builder.ConnectionString);
conn.Open();
Console.WriteLine(conn.State);
```
I will test with a remote on premises server later today.

**Comment by ctrlaltdan** [comment]
Also seeing the same. Downgrading to `5.2.0-preview5.24024.3` resolved the issue for me.

Additional debug info if it's helpful.

```
Framework: .NET 8.0.2 
Runtime: linux-musl-x64
Image: Alpine Linux v3.19
Using: Microsoft.EntityFrameworkCore.SqlServer:8.0.2
Connected using an Azure SQL Failover group, via Entity Framework. 
```

**Comment by David-Engel** [comment]
If you are on Linux/macOS and specify both port and instance name in the connection string (like server,12345\instance), that might be the source issue. There appears to have been a regression in 5.2.0 on non-Windows where it isn't ignoring the instance name when both it and the port are specified.

**Comment by sturledahl** [comment]
Some obervation from me hoping it helps investigation:

We started getting this problem in alpine for a test that starts multiple threads connecting some same database in parallel. Other tests work fine. Our connection strings do not specify instance or port. Tried adding some delays in the code inside the different tasks to affect timing and then we got a different error instead of the one mentioned in this ticket:

`System.InvalidOperationException: Timeout expired.  The timeout period elapsed prior to obtaining a connection from the pool.  This may have occurred because all pooled connections were in use and max pool size was reached.
12:25:02        at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)`

Here's how threads are created in the test:
`            var tasks = schedulers.Select(s => new TaskFactory().StartNew(s.Start)).ToList();
            foreach (var t in tasks)
                t.Wait();
`

Reverting back to v5.1.5 fixed this, so we are not updating to v5.2.0 until we know more.

**Comment by alex-jitbit** [comment]
I can confirm that we too experience this error under a heavy load with multiple threads (not sure if this is the culprit)

**Comment by JRahnama** [comment]
can you guys test with this package and see if the issue is resolved? just change the extension to `nupkg` and should be good for testing.
[Microsoft.Data.SqlClient.6.0.0-pull.106802.zip](https://github.com/dotnet/SqlClient/files/14540821/Microsoft.Data.SqlClient.6.0.0-pull.106802.zip)

**Comment by sturledahl** [comment]
@JRahnama Any chance you could add it to nuget.org so our build/test system can find it?

**Comment by JRahnama** [comment]
@sturledahl this package is not officially signed and is not suitable for production use. I just wanted to confirm that the fix has resolved the issue for users before proceeding with a hotfix release.

**Comment by PaulVrugt** [comment]
any update on this?

**Comment by JRahnama** [comment]
> any update on this?

Were you able to test with the sample package?

> can you guys test with this package and see if the issue is resolved? just change the extension to `nupkg` and should be good for testing. [Microsoft.Data.SqlClient.6.0.0-pull.106802.zip](https://github.com/dotnet/SqlClient/files/14540821/Microsoft.Data.SqlClient.6.0.0-pull.106802.zip)

**Comment by PaulVrugt** [comment]
@JRahnama well no. We haven't updated to 5.2 yet because of this issue. We are currently using version 5.1.5 and running into https://github.com/dotnet/SqlClient/issues/449, but it is only happening in our production environment (with a lot of traffic) and even there only once every few weeks. We have no controlled environment to test this. Maybe @alex-jitbit has a way to reproduce it and see if the 6.0.0 version resolves it

**Comment by mosesnnewman** [comment]
Same issue on Windows!

**Comment by JRahnama** [comment]
@alex-jitbit is it possible to test with 5.2.0-preview2 and 5.2.0-preview5 versions to identify what changed caused the issue?

**Comment by ABAG603** [comment]
We have same issue as people above after upgrading to the 5.2.0. Issue is reproducible from either Alpine containers or VM with Amazon Linux 2023. SQL Server is running on Windows Server and connection string contains named instance and port.

@JRahnama I've tested with different versions and here's the outcome:
- 5.2.0-preview2.23159.1 - connection can be successfully opened
- 5.2.0-preview5.24024.3 - connection can be successfully opened
- 5.2.0 - opening connection fails with error below

```
Microsoft.Data.SqlClient.SqlException: A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct
and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 26 - Error Locating Server/Instance Specified)
     System.Net.Sockets.SocketException: Success
       at int Microsoft.Data.SqlClient.SNI.SSRP.GetPortByInstanceName(string browserHostName, string instanceName, TimeoutTimer timeout, bool allIPsInParallel, SqlConnectionIPAddressPreference ipPreference)
       at SNITCPHandle Microsoft.Data.SqlClient.SNI.SNIProxy.CreateTcpHandle(DataSource details, TimeoutTimer timeout, bool parallel, SqlConnectionIPAddressPreference ipPreference, string cachedFQDN, ref SQLDNSInfo pendingDNSInfo,
          bool tlsFirst, string hostNameInCertificate, string serverCertificateFilename)
  at void Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, bool breakConnection, Action<Action> wrapCloseInAction)
  at void Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, SqlCommand command, bool callerHasConnectionLock, bool asyncClose)
  at void Microsoft.Data.SqlClient.TdsParser.Connect(ServerInfo serverInfo, SqlInternalConnectionTds connHandler, TimeoutTimer timeout, SqlConnectionString connectionOptions, bool withFailover)
  at void Microsoft.Data.SqlClient.SqlInternalConnectionTds.AttemptOneLogin(ServerInfo serverInfo, string newPassword, SecureString newSecurePassword, TimeoutTimer timeout, bool withFailover)
  at void Microsoft.Data.SqlClient.SqlInternalConnectionTds.LoginNoFailover(ServerInfo serverInfo, string newPassword, SecureString newSecurePassword, bool redirectedUserInstance, SqlConnectionString connectionOptions,
     SqlCredential credential, TimeoutTimer timeout)
  at void Microsoft.Data.SqlClient.SqlInternalConnectionTds.OpenLoginEnlist(TimeoutTimer timeout, SqlConnectionString connectionOptions, SqlCredential credential, string newPassword, SecureString newSecurePassword, bool
     redirectedUserInstance)
  at Microsoft.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, object providerInfo, string newPassword, SecureString
     newSecurePassword, bool redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, bool applyTransientFaultHandling, string accessToken, DbConnectionPool pool,
     Func<SqlAuthenticationParameters, CancellationToken, Task<SqlAuthenticationToken>> accessTokenCallback)
  at DbConnectionInternal Microsoft.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection,
     DbConnectionOptions userOptions)
  at DbConnectionInternal Microsoft.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool, DbConnection owningObject, DbConnectionOptions options, DbConnectionPoolKey poolKey, DbConnectionOptions
     userOptions)
  at DbConnectionInternal Microsoft.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
  at DbConnectionInternal Microsoft.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection)
  at bool Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, uint waitForMultipleObjectsTimeout, bool allowCreate, bool onlyOneCheckConnection, DbConnectionOptions userOptions, out
     DbConnectionInternal connection)
  at bool Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource<DbConnectionInternal> retry, DbConnectionOptions userOptions, out DbConnectionInternal connection)
  at bool Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource<DbConnectionInternal> retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, out
     DbConnectionInternal connection)
  at bool Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource<DbConnectionInternal> retry, DbConnectionOptions
     userOptions)
  at bool Microsoft.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource<DbConnectionInternal> retry, DbConnectionOptions userOptions)
  at bool Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource<DbConnectionInternal> retry, SqlConnectionOverrides overrides)
  at void Microsoft.Data.SqlClient.SqlConnection.Open(SqlConnectionOverrides overrides)
  at void Microsoft.Data.SqlClient.SqlConnection.Open()
  at void Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerConnection.OpenDbConnection(bool errorsExpected)
  at void Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenInternal(bool errorsExpected)
  at bool Microsoft.EntityFrameworkCore.Storage.RelationalConnection.Open(bool errorsExpected)
  at bool Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.<>c.<OpenConnection>b__22_0(DatabaseFacade database)
  at TResult Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.<>c__DisplayClass12_0`2.<Execute>b__0(DbContext _, TState s)
  at TResult Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute<TState,TResult>(TState state, Func<DbContext, TState, TResult> operation, Func<DbContext, TState, ExecutionResult<TResult>>
     verifySucceeded)
  at TResult Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute<TState,TResult>(IExecutionStrategy strategy, TState state, Func<TState, TResult> operation, Func<TState, ExecutionResult<TResult>> verifySucceeded)
  at void Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.OpenConnection(DatabaseFacade databaseFacade)
  at int TestConnectionCommand.Execute(CommandContext context)
```

**Comment by JRahnama** [comment]
@ABAG603 Fix is merged in the main branch by PR #2395. Hotfix release v5.2.1 is planned, but date yet TBD.

Closing the issue as fix is merged and will be available by next hotfix release.

**Comment by hssamany** [comment]
Unfortunatly, I couldn't wait till release, But I found out that, one other cause of this error can be due to the compatibility level of your sql-server. For example Linq-Queries with a collection filter like so `query.Where(p => (new List<string> { "XX","YY"})).Contains(p.MyCode));` resulted into an SQL like:
 
```
SELECT [t].[Id]
WHERE [t].[MyCode] IN (
    SELECT [c].[value]
    FROM OPENJSON(@__codes_0) WITH ([value] nvarchar(50) '$') AS [c]
)
ORDER BY [t].[Id]
```.

Then I had to raise my Database Compatibility level to "150" in order to run the "OPENJSON"-function. But I assume it should also work under level "130".

**Comment by PaulVrugt** [comment]
@JRahnama any update on a release date for 5.2.1 yet? It's been a couple of weeks now and still no hotfix

**Comment by alex-jitbit** [comment]
NOT FIXED attn @JRahnama 

The issue is not fixed in 5.2.1

I'm getting the same error under high load:

```
Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)
 ---> System.TimeoutException: The socket couldn't connect during the expected 14999 remaining time.
   at Microsoft.Data.SqlClient.SNI.SNITCPHandle.Connect(String serverName, Int32 port, TimeoutTimer timeout, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo)
   at Microsoft.Data.SqlClient.SNI.SNITCPHandle..ctor(String serverName, Int32 port, TimeoutTimer timeout, Boolean parallel, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo, Boolean tlsFirst, String hostNameInCertificate, String serverCertificateFilename)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry, SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open(SqlConnectionOverrides overrides)
   ```
   
   Reverting to 5.1.5 fixes the issue

**Comment by alex-jitbit** [comment]
Please reopen the issue

**Comment by JRahnama** [comment]
@alex-jitbit Can you confirm that the issue is not happening with 5.2.0-preview5.24024.3? I saw some other users have confirmed it and that was regarding a regression we addressed in 5.2.1. Seems like in your case is a bit different.

**Comment by JRahnama** [comment]
Also we are going to need a simple repro application for further investigation.

**Comment by David-Engel** [comment]
@alex-jitbit 
Does increasing the connect timeout in your connection string solve the issue? The reason I ask is that #2098 in 5.2-preview improved respecting of connection timeout during connect on the exact path your exception is occurring and the message indicates the default timeout of 15 seconds has elapsed when the exception is thrown. I'm wondering if 5.1 was simply taking longer than the connect timeout to connect under load but succeeding anyway.

**Comment by alex-jitbit** [comment]
@JRahnama repro here: https://gist.github.com/alex-jitbit/1eca9a1f014e036691bdc35cd852c726 the bug is even reproducable when running on Windows under WSL2 however the error is slightly different in that case (see repro description)

**Comment by apxltd** [comment]
I can confirm that this is indeed broken in 5.2.1 as well, and **actually worse than 5.2.0**. This is trivial to reproduce, basically do exactly what @alex-jitbit -- just open a bunch of connections to simulate what a real application experiencing traffic would do. 

In 5.2.0, SqlClient crashes immediately with that same. When we use 5.2.1, there seems to be some kind of timeout mechanism that makes the cashes take forever, and effectively freezes up the whole application. 

Stack trace seems to be the same:

```
Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)
---> System.TimeoutException: The socket couldn't connect during the expected 14990 remaining time.
at Microsoft.Data.SqlClient.SNI.SNITCPHandle.Connect(String serverName, Int32 port, TimeoutTimer timeout, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo)
at Microsoft.Data.SqlClient.SNI.SNITCPHandle..ctor(String serverName, Int32 port, TimeoutTimer timeout, Boolean parallel, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo, Boolean tlsFirst, String hostNameInCertificate, String serverCertificateFilename)
at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
at Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry, SqlConnectionOverrides overrides)
at Microsoft.Data.SqlClient.SqlConnection.InternalOpenAsync(CancellationToken cancellationToken)
```

**Comment by rodion-m** [comment]
This issue should be pinned.

**Comment by JRahnama** [comment]
Is it possible to test with an unsigned package to see if the issue is resolved? Just change the extension to .nupkg and consume the package in your project.
[Microsoft.Data.SqlClient.6.0.0-pull.118801.zip](https://github.com/user-attachments/files/16117464/Microsoft.Data.SqlClient.6.0.0-pull.118801.zip)

**Comment by alex-jitbit** [comment]
@JRahnama just tested with the exact repro I provided, same error on 40 concurrent connections.

**Comment by altso** [comment]
Because of [this issue](https://github.com/dotnet/SqlClient/issues/1108), staying on version 5.1.5 also prevents you from automatically receiving vulnerability fixes for `Azure.Identity` and `Microsoft.Identity.Client` available in the later versions of `Microsoft.Data.SqlClient`.

**Comment by PaulVrugt** [comment]
I don't understand why this issue isn't receiving more traction. The few people responding to this cannot be the only persons running into this issue. The package simply doesn't seem to work properly when having a lot of connections.

**Comment by ErikEJ** [comment]
@PaulVrugt  a lot of people run on Windows or stay on 5.1.5

**Comment by andrew-from-toronto** [comment]
We are seeing this issue in production as well, although it seems pretty random when it occurs. We use Entity Framework Extensions and notice calls from that library seem to run into it more. We are pretty stuck here since we upgraded to dotnet8 and this started happening, would love to see this fixed.

**Comment by saithis** [comment]
We encountered this issue in our testing cluster when trying to update, so we just didn't.

We use Azure k8s with managed azure sql and I don't think we do anything special/weird in our apps. The only thing maybe being, that we have multiple ef core db contexts to different databases.

Even with low usage (a single person clicking around) we instantly run into problems.

**Comment by PaulVrugt** [comment]
@JRahnama this issue is still tagged as "waiting for customer", but apparently there are simple reproduction steps. What more info is required for this?

**Comment by JRahnama** [comment]
> @JRahnama this issue is still tagged as "waiting for customer", but apparently there are simple reproduction steps. What more info is required for this?

that was an oversight. We will investigate the issue and will get back here.

**Comment by JRahnama** [comment]
This issue is the result of improving time outs which happened on 5.2.0-preview4 in PR #2098 which could be a duplicate of issue #422. I can suggest as a workaround to manage available threads to your application by setting below:

```C#
ThreadPool.GetMaxThreads(out var workerThreads, out var completionThreads);
ThreadPool.SetMinThreads(<number of parallel/concurent connections>, completionThreads);
```

Previously we were not checking for the time out and connections were working as expected, but by adding/honoring timeout the actual issue has surfaced. Work is in progress to address the root cause by redesigning the driver in SqlClientX. 

Let me know if proposed workaround has done the job for now, until the fix is out with [SqlClientX](https://github.com/dotnet/SqlClient/tree/feat/sqlclientx),

**Comment by ErikEJ** [comment]
@JRahnama Would increasing the connection and command timeouts via the connection string have an effect?

**Comment by altso** [comment]
@JRahnama what's SqlClientX? Do you have an ETA for the fix to land?

**Comment by JRahnama** [comment]
> @JRahnama Would increasing the connection and command timeouts via the connection string have an effect

Increasing ConnecTimeout to 60 seconds has also solved the issue for the provided repro.

**Comment by JRahnama** [comment]
> @JRahnama what's SqlClientX? Do you have an ETA for the fix to land?

you can have a look at the broader picture in [this discussion](https://github.com/dotnet/SqlClient/discussions/2603#discussioncomment-9868402).

**Comment by alex-jitbit** [comment]
Please **DO NOT** use the proposed "workaround" in an ASP.NET Core app, you will ruin your performance. More CPU time dequeuing work, more context switching, more garbage collection, more memory usage., etc etc.

MS's own docs caution against that.

**Comment by JRahnama** [comment]
> Please **DO NOT** use the proposed ["workaround"](https://github.com/dotnet/SqlClient/issues/2378#issuecomment-2221231526) in an ASP.NET Core app, you will ruin your performance. More CPU time dequeuing work, more context switching, more garbage collection, more memory usage., etc etc.
> 
> MS's own docs caution against that.

Then increasing ConnectTimeout would be your only solution.

**Comment by apxltd** [comment]
@JRahnama I'm sorry to be so negative here, but your responses are completely inadequate and totally unacceptable. They demonstrate a lack of empathy and acknowledgement of the problems your team has caused.

You need to keep in mind that Microsoft's customers pay a preposterous amount of money for SQL Server and your team's untested changes have made it effectively unusable in Linux. I really can't tell if your team doesn't understand this... or they simply don't care.

If Microsoft's plan is for us to stop using SQL Server -- which would make sense, when they can charge the [same amount of money to host a database that someone else maintains for free](https://azure.microsoft.com/en-us/products/postgresql/) -- then just please just say so. While that's a lot of work for us, at least PostgreSQL's team of volunteer contributors demonstrates more more professionalism that your team has been.

Otherwise, at the very least, please run your replies through Bing AI to at least feign some kind of empathy and learn what key points you need to say when responding to a showstopper bug from your customers.

**Comment by JRahnama** [comment]
@apxltd Thank you for your feedback. We appreciate you taking the time to share your concerns with us.

**Comment by apxltd** [comment]
@JRahnama hmmm okay okay.... maybe switch to Chat GPT instead of Bing AI 😉

Here's what it suggested when I prompted "An annoyed customer responded with this. Please help me craft a short responses."

> Thank you for bringing this to our attention. I apologize for the frustration and inconvenience you’ve experienced. Your feedback is invaluable, and I want to assure you that we take these issues very seriously.
> 
> We understand the significant investment you’ve made in SQL Server, and it’s clear that the recent changes have negatively impacted your operations. This is not the level of service we aim to provide, and we are committed to resolving this issue promptly.
> 
> Our team is currently investigating the root cause of the problem and will work diligently to find a solution. We will keep you updated with our progress and ensure that any changes are thoroughly tested moving forward.
> 
> Your patience and understanding are greatly appreciated. If you have any further concerns or require immediate assistance, please do not hesitate to reach out.

You see, a high-quality GPT response like that gives you the impression that you aren't being passive-aggressive. And it *almost* gives the impression that you care about the problems you have caused for your customers.

**Comment by JRahnama** [comment]
@alex-jitbit if increasing the ConnectTimeout is not a preferred solution, I may have a solution for this issue until SqlClientX is out for good. Give me some time to test my proposed solution and discuss that internally with other team members, in case of positive outcome I will give you another nuget to test with.

**Comment by JRahnama** [comment]
> So I'm trying to find the most suitable workaround for our application that works properly in production.
> 
> am I to understand that if I increase the connect timeout parameter the issue does not occur? Will it then still take a lot of time to connect each time. Or does it only take a lot of time until enough threads are created to handle all connections, and will it then be fast until the application is restarted?
> 
> I also don't really understand the resistance of setting min threads. If apparently your application needs that amount of threads in production workloads, what is the issue in setting it so it starts with enough threads?
> 
> Nb: let's please try to keep this discussion productive. I too would like to stress this is a pressing issue for us, but let's not make this a toxic discussion

In my testing with 50 concurrent connections, I did not notice any specific performance issues. This is based on observation rather than benchmark testing. It worked pretty well when I increased the timeout to 60 seconds. This provided enough time for the socket to connect. I think that would be the optimal workaround for now. However, as I mentioned before, give me some time to test my findings, and I will get back to you.

**Comment by BlythMeister** [comment]
Is there scope to create a 5.1.6 which is 5.1.5 with the dependencies on Azure.Identity and Microsoft.Identity.Client updated to patches with CVE-2024-35255 fixes applied.
I fully understand find a "proper" solution may take time, but forcing users connecting to SQL server and running on linux into unproven workarounds on PRD systems to be able to patch against such a serious vulnerability in another Microsoft library is completely unacceptable.

A patch of 5.1.5 which updates the transient dependencies with vulnerabilities is (in my opinion) essential.  It is now nearly 2 months since these transient dependencies were updated to fix the CVE, but it's impossible to get these updates.

**Comment by PaulVrugt** [comment]
> > So I'm trying to find the most suitable workaround for our application that works properly in production.
> > am I to understand that if I increase the connect timeout parameter the issue does not occur? Will it then still take a lot of time to connect each time. Or does it only take a lot of time until enough threads are created to handle all connections, and will it then be fast until the application is restarted?
> > I also don't really understand the resistance of setting min threads. If apparently your application needs that amount of threads in production workloads, what is the issue in setting it so it starts with enough threads?
> > Nb: let's please try to keep this discussion productive. I too would like to stress this is a pressing issue for us, but let's not make this a toxic discussion
> 
> In my testing with 50 concurrent connections, I did not notice any specific performance issues. This is based on observation rather than benchmark testing. It worked pretty well when I increased the timeout to 60 seconds. This provided enough time for the socket to connect. I think that would be the optimal workaround for now. However, as I mentioned before, give me some time to test my findings, and I will get back to you.

Any update on this @JRahnama ?

**Comment by JRahnama** [comment]
@PaulVrugt Unfortunately, my testing didn't yield positive results. I've explored various solutions within the current design, but none were effective and each had its own drawback.

In the meantime, here are some workarounds I've found:

1. Increase ConnectTimeout: This can help by providing more time for socket connection.
2. Set Minimum Available Threads: Adjusting the minimum available threads in your application might help, although it could impact performance.
3. Disable Pooling: Turning off pooling creates a new connection each time, which is not ideal as it’s quite resource-intensive.

I understand that the issue is frustrating, but I want to assure you that we're doing our best to resolve it as quickly as possible. Our team is actively working on redesigning the pool and concurrent connections, and it's progressing well.

**Comment by BlythMeister** [comment]
> @PaulVrugt Unfortunately, my testing didn't yield positive results. I've explored various solutions within the current design, but none were effective and each had its own drawback.
> 
> In the meantime, here are some workarounds I've found:
> 
> 1. Increase ConnectTimeout: This can help by providing more time for socket connection.
> 2. Set Minimum Available Threads: Adjusting the minimum available threads in your application might help, although it could impact performance.
> 3. Disable Pooling: Turning off pooling creates a new connection each time, which is not ideal as it’s quite resource-intensive.
> 
> I understand that the issue is frustrating, but I want to assure you that we're doing our best to resolve it as quickly as possible. Our team is actively working on redesigning the pool and concurrent connections, and it's progressing well.

Whilst i fully appreciate your working on a proper fix for this issue.
Is there really no scope to release a hotfix to 5.1.5 with the dependencies updated as per my comment - https://github.com/dotnet/SqlClient/issues/2378#issuecomment-2247689996 ??

**Comment by ErikEJ** [comment]
@BlythMeister Yes, 5.1.6 is being worked on

**Comment by noragarkalne** [comment]
Hello, we have .NET 8 and app hosted on in azure
`Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)
 ---> System.TimeoutException: The socket couldn't connect during the expected 14999 remaining time.
   at Microsoft.Data.SqlClient.SNI.SNITCPHandle.Connect(String serverName, Int32 port, TimeoutTimer timeout, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo)
   at Microsoft.Data.SqlClient.SNI.SNITCPHandle..ctor(String serverName, Int32 port, TimeoutTimer timeout, Boolean parallel, SqlConnectionIPAddressPreference ipPreference, String cachedFQDN, SQLDNSInfo& pendingDNSInfo, Boolean tlsFirst, String hostNameInCertificate, String serverCertificateFilename)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
`
   Do I understand right that this is known issue that is caused by this package or I have some wrong setup somewhere? I have currently 5.2.1 version in my project

**Comment by cbi-at-varian** [comment]
The provided workarounds aren't mitigating the issue completely in our case. The only "solution" is actually to downgrade to 5.1.5

We're load testing our service with k6 using the [ramping-vus](https://k6.io/docs/using-k6/scenarios/executors/ramping-vus/) executor. We're ramping it up to 1000 vus and let the test run for 3min. For one of our scenario we tested using different connection timeouts with the 5.2.1 version as suggested workaround 1)

| setup details                                    | average request time | minimum request time | median request time | max request time | failed calls out of max    |
|--------------------------------------------------|----------------------|----------------------|---------------------|------------------|----------------------------|
| SqlClient.5.2.1; default Connection Timeout (15) | 356.52ms             | 1.18ms               | 346.64ms            | 15.51s           | failed 1.41% (1605/112097) |
| SqlClient.5.2.1; Connection Timeout=30           | 361.04ms             | 1.21ms               | 344.56ms            | 30.66s           | failed 1.17% (1418/119465) |
| SqlClient.5.2.1; Connection Timeout=60           | 358.13ms             | 1.19ms               | 347.49ms            | 1m0s             | failed 0.27% (316/116617)  |
| SqlClient.5.1.5; default Connection Timeout (15) | 384.03m              | 1.08ms               | 367.17ms            | 1.45s            | failed 0% (0/119816 )      |

The higher we put the timeout the better it went. Going to 180s we got zero errors, but we started to get client timeouts. Which is not acceptable as well. 

The suggested workaround 2) to increase the Threadpool minimum size did not provide any better results. Honestly, we didn't try out the suggested workaround 3) as it sounds just too scary..


So please a proper fix would be really appreciated. Is there in the meantime an ETA for the fix?

**Comment by ahouben** [comment]
> @BlythMeister Yes, 5.1.6 is being worked on

@ErikEJ @JRahnama would you know more about an ETA for when the connection pooling/handling is fixed and released in a 5.1.6 (hotfix) or 5.2.2/3 (on top of HEAD) version ?

From https://www.nuget.org/packages/Microsoft.Data.SqlClient...
![image](https://github.com/user-attachments/assets/1f582c5f-e3b7-483a-b85e-adbb3b563b41)

**Comment by robs** [comment]
So, have I read this wrong or does the latest stable release in nuget have this issue?

**Comment by ahouben** [comment]
@robs The latest version 5.2.1 on nuget still has this issue

**Comment by JonasJes** [comment]
> Oh wow, surely the "stable" flag should be removed then?
> 
> We've not updated to 5.2.x yet but we're running Linux clients to Windows MS SQL.
> 
> Is everyone just waiting in the hope there's a 5.1.x release that has the updated dependencies?

@cbi-at-varian Great work showing what we are experiencing

**Comment by JRahnama** [comment]
> Is everyone just waiting in the hope there's a 5.1.x release that has the updated dependencies?

5.1.x will be released soon ( in a week or so) with updated dependencies.

**Comment by Markeli** [comment]
And what about 5.2.x? Will it also be released soon?

**Comment by JRahnama** [comment]
Could someone please test the package below to see if the issue is resolved? Simply change the extension to `nupkg`. Note that this is an unsigned package and is intended for testing purposes only.
[Microsoft.Data.SqlClient.6.0.0-pull.122701.zip](https://github.com/user-attachments/files/16619846/Microsoft.Data.SqlClient.6.0.0-pull.122701.zip)

**Comment by JRahnama** [comment]
> And what about 5.2.x? Will it also be released soon?

If the fix works it will backported to 5.2 and the hot fix will be released in second half of August.

**Comment by JRahnama** [comment]
The issue was at [TCPHanlde.Connect](https://github.com/dotnet/SqlClient/blob/d3658edfcda0d7eecabc09afb5459bb3f79e6290/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNITcpHandle.cs#L440) function. Having those async tasks in the middle of the primary sync flow was causing thread pool starvation under load. The proposed solution was to remove that as we are using socket.Select() later with the timeout to honor the timeout. Which was also mentioned at [here](https://github.com/dotnet/SqlClient/issues/826#issuecomment-736224118) by @roji

**Comment by cbi-at-varian** [comment]
> Could someone please test the package below to see if the issue is resolved? Simply change the extension to `nupkg`. Note that this is an unsigned package and is intended for testing purposes only. [Microsoft.Data.SqlClient.6.0.0-pull.122701.zip](https://github.com/user-attachments/files/16619846/Microsoft.Data.SqlClient.6.0.0-pull.122701.zip)
> 
> Update: Our test pipelines passed with the changes, and the provided repro is working well with this set of modifications. The fix appears promising. I’ll wait for the impacted users' test results before proceeding with the PR. Also I will look into adding some tests if possible to prevent this from happening in future. Thank you all for your patience. Also thanks to @David-Engel for pointing out the root cause and solution.

I integrated the package provided by @JRahnama  and run the load tests. It seems to fix the issue, thanks for that! 

Here the numbers to compare against:
| setup details                                    | average request time | minimum request time | median request time | max request time | failed calls out of max    |
|--------------------------------------------------|----------------------|----------------------|---------------------|------------------|----------------------------|
| SqlClient.5.1.5; default Connection Timeout (15) | 398.72m              | 1.19ms               | 376.23ms            | 1.49s            | failed 0% (0/226103)      |
| SqlClient.6.0.0.pull; default Connection Timeout (15) | 434.98ms             | 1.3ms               | 401.33ms            | 1.51s           | failed 0% (0/207359) |

Please note the numbers do not imply a certain quality of the fix, the load test runs aren't deterministic. Important for me is that we keep the expected thresholds and that there are no failures. It feels the fix is slightly slower, but only a benchmark would reveal that.

**Comment by saurabh500** [comment]
@cbi-at-varian what does the `average request time` column measure? The average time taken to open a connection?

**Comment by JRahnama** [comment]
Here are benchmarks after the fix:

BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4037/23H2/2023Update/SunValley3)
13th Gen Intel Core i7-1365U, 1 CPU, 12 logical and 10 physical cores
.NET SDK 8.0.400
  [Host]   : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2
  .NET 8.0 : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2

Job=.NET 8.0  Runtime=.NET 8.0

| Method                 | Mean     | Error   | StdDev  | Ratio | RatioSD |
|----------------------- |---------:|--------:|--------:|------:|--------:|
| OpenSqlConnectionAsync | 172.3 ms | 3.05 ms | 2.55 ms |  1.00 |    0.02 |



below are results from  **5.1.5:**:

BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4037/23H2/2023Update/SunValley3)
13th Gen Intel Core i7-1365U, 1 CPU, 12 logical and 10 physical cores
.NET SDK 8.0.400
  [Host]   : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2
  .NET 8.0 : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2

Job=.NET 8.0  Runtime=.NET 8.0

| Method                 | Mean     | Error   | StdDev  | Ratio | RatioSD |
|----------------------- |---------:|--------:|--------:|------:|--------:|
| OpenSqlConnectionAsync | 177.0 ms | 3.52 ms | 4.58 ms |  1.00 |    0.04 |

**Comment by JRahnama** [comment]
There are a couple of PRs to be bacported to 5.2 and the we will proceed with the release of hotfix 5.2.2

**Comment by cbi-at-varian** [comment]
> @cbi-at-varian what does the `average request time` column measure? The average time taken to open a connection?

@saurabh500 the measurements are on the client side of the load tests. Those numbers therefore include the web stack, some small business logic including accessing the database.

**Comment by bdschaap** [comment]
> Could someone please test the package below to see if the issue is resolved? Simply change the extension to `nupkg`. Note that this is an unsigned package and is intended for testing purposes only. [Microsoft.Data.SqlClient.6.0.0-pull.122701.zip](https://github.com/user-attachments/files/16619846/Microsoft.Data.SqlClient.6.0.0-pull.122701.zip)
> 
> Update: Our test pipelines passed with the changes, and the provided repro is working well with this set of modifications. The fix appears promising. I’ll wait for the impacted users' test results before proceeding with the PR. Also I will look into adding some tests if possible to prevent this from happening in future. Thank you all for your patience. Also thanks to @David-Engel for pointing out the root cause and solution.

Can a preview nuget be pushed?

**Comment by JRahnama** [comment]
> Can a preview nuget be pushed?

M.D.SqlClient v6.0.0-preview1 and hotfix release v5.2.2 are awaiting to be released in a day or two

**Comment by JRahnama** [comment]
Hotfix [v5.2.2](https://www.nuget.org/packages/Microsoft.Data.SqlClient/5.2.2) is released.

**Comment by alex-jitbit** [comment]
We're still getting this under very heavy load.

```
Microsoft.Data.SqlClient.SqlException (0x80131904): Connection Timeout Expired.  The timeout period elapsed while attempting to consume the pre-login handshake acknowledgement.  This could be because the pre-login handshake failed or the server was unable to respond back in time.  The duration spent while attempting to connect to this server was - [Pre-Login] initialization=0; handshake=15006;
```

Rolling back to 5.1.5 fixes the issue

**Comment by sturledahl** [comment]
@alex-jitbit or @patrickg-hchb or someone else, do you still see this problem with latest version v6.0.1? We have locked all our apps to v5.1.5 until now due to this issue, but want to go to latest. Just not sure if we should risk it..

**Comment by alex-jitbit** [comment]
@sturledahl we are on 5.2.2 all good. Looks safe.

P.S. Had to rewrite hot path code to ease the load on connection pool (tighten "open/close" timing and eliminate multiple connections when several methods call each other in a chain).
